packages/labs/compiler/src/lib/ast-fragments.ts

```diff
@@ -7,14 +7,175 @@
 import ts from 'typescript';
 
 export const PartType = {
+  ATTRIBUTE: 1,
   CHILD: 2,
+  PROPERTY: 3,
+  BOOLEAN_ATTRIBUTE: 4,
+  EVENT: 5,
+  ELEMENT: 6,
+  COMMENT_PART: 7,
 } as const;
 
 export type PartType = (typeof PartType)[keyof typeof PartType];
 
-export type TemplatePart = {
-  type: typeof PartType.CHILD;
-  index: number;
+export type TemplatePart =
+  | {
+      type:
+        | typeof PartType.CHILD
+        | typeof PartType.COMMENT_PART
+        | typeof PartType.ELEMENT;
+      index: number;
+    }
+  | {
+      type: typeof PartType.ATTRIBUTE;
+      index: number;
+      name: string;
+      strings: Array<string>;
+      ctorType: PartType;
+    };
+
+const attributePartConstructors = {
+  [PartType.ATTRIBUTE]: 'AttributePart',
+  [PartType.PROPERTY]: 'PropertyPart',
+  [PartType.BOOLEAN_ATTRIBUTE]: 'BooleanAttributePart',
+  [PartType.EVENT]: 'EventPart',
+} as const;
```

Comment:
Consider separating out the attribute kind as its own type, because `ctorType` can't be any PartType, it's just one of the attribute kinds. Likewise, maybe we should remove the values that can only be attribute kinds from PartType

```suggestion
export const AttributeKind = {
  ATTRIBUTE: 1,
  PROPERTY: 3,
  BOOLEAN_ATTRIBUTE: 4,
  EVENT: 5,
} as const;

export type AttributeKind = (typeof AttributeKind)[keyof typeof AttributeKind];

export type TemplatePart =
  | {
      type:
        | typeof PartType.CHILD
        | typeof PartType.COMMENT_PART
        | typeof PartType.ELEMENT;
      index: number;
    }
  | {
      type: typeof PartType.ATTRIBUTE;
      index: number;
      name: string;
      strings: Array<string>;
      ctorType: AttributeKind;
    };

const attributePartConstructors = {
  [AttributeKind.ATTRIBUTE]: 'AttributePart',
  [AttributeKind.PROPERTY]: 'PropertyPart',
  [AttributeKind.BOOLEAN_ATTRIBUTE]: 'BooleanAttributePart',
  [AttributeKind.EVENT]: 'EventPart',
} as const;
```